components/core/src/clp/StreamingReader.cpp

```diff
@@ -0,0 +1,353 @@
+    if (0 != offset) {
+        std::string const range{"Range: bytes=" + std::to_string(offset) + "-"};
+        request_header = curl_slist_append(request_header, range.c_str());
+    }
+    if (disable_caching) {
+        request_header = curl_slist_append(request_header, "Cache-Control: no-cache");
+        request_header = curl_slist_append(request_header, "Pragma: no-cache");
+    }
+    if (nullptr != request_header) {
+        curl_easy_setopt(curl_handle, CURLOPT_HTTPHEADER, request_header);
+    }
+
+    auto const retval{curl_easy_perform(curl_handle)};
+    reader.commit_fetching_buffer();
+    reader.m_status_code = (0 != retval) ? StatusCode::Failed : StatusCode::Finished;
+    reader.m_curl_return_code = retval;
+
+    curl_easy_cleanup(curl_handle);
+    reader.m_transfer_terminated.store(true);
+}
+
+auto StreamingReader::progress_callback(
+        void* ptr,
+        [[maybe_unused]] curl_off_t dltotal,
+        [[maybe_unused]] curl_off_t dlnow,
+        [[maybe_unused]] curl_off_t ultotal,
+        [[maybe_unused]] curl_off_t ulnow
+) -> int {
+    auto const* reader{get_reader(ptr)};
+    return reader->is_transfer_aborted() ? 1 : 0;
+}
+
+auto StreamingReader::write_callback(char* ptr, size_t size, size_t nmemb, void* reader_ptr)
+        -> size_t {
+    auto const num_total_bytes{size * nmemb};
+    try {
+        StreamingReader::BufferView input_buffer{ptr, num_total_bytes};
+        auto* reader{get_reader(reader_ptr)};
+        while (false == input_buffer.empty()) {
+            auto const num_bytes_left{input_buffer.size()};
+            StreamingReader::BufferView fetching_buffer;
+            if (false == reader->get_buffer_to_fetch(fetching_buffer)) {
+                return 0;
+            }
+            auto const num_bytes_to_fetch{
+                    num_bytes_left > fetching_buffer.size() ? fetching_buffer.size()
+                                                            : num_bytes_left
+            };
+            memcpy(fetching_buffer.data(), input_buffer.data(), num_bytes_to_fetch);
+            input_buffer = input_buffer.last(num_bytes_left - num_bytes_to_fetch);
+            reader->commit_fetching(num_bytes_to_fetch);
+        }
+    } catch (TraceableException const& ex) {
+        // TODO: add logging to track the exceptions.
+        return 0;
+    }
+    return num_total_bytes;
+}
+
+auto StreamingReader::open(std::string_view src_url, size_t offset, bool disable_caching) -> void {
+    if (false == m_initialized) {
+        throw OperationFailed(ErrorCode_NotReady, __FILE__, __LINE__);
+    }
+    if (StatusCode::NotInit != m_status_code) {
+        throw OperationFailed(ErrorCode_NotReady, __FILE__, __LINE__);
+    }
+    m_src_url = src_url;
+    m_transfer_thread = std::make_unique<std::thread>(
+            transfer_thread_entry,
+            std::ref(*this),
+            offset,
+            disable_caching
+    );
+    m_status_code = StatusCode::InProgress;
```

Comment:
hmmm, we don't have any assumptions that this interface will be used across different threads. At the beginning of the function, we checked that the status must be `NotInit` otherwise `open` should fail. If no other threads will try to open the interface concurrently, it should be ok to set the status to `InProgress` here before the fetching thread gets executed.

Reply:
`m_status_code` is used in both the main thread and the fetching thread. There's no guarantee that this line is executed before the fetching thread gets executed (though in reality it usually is). It is possible that the fetching thread sets it to `StatusCode::Failed` and the main thread sets it to `StatusCode::InProgress`, which leaves it in a wrong status. But theres should be no problem of the current implementation since `terminate_current_transfer` can anyway handle the status correctly.

Reply:
Sorry, I was wrong. I forgot that `std::thread` starts executing after initialization. In this case, do u think it's better move this line before the thread creation?

Reply:
Probably we can set it in `transfer_thread_entry`?

Reply:
> Sorry, I was wrong. I forgot that `std::thread` starts executing after initialization. In this case, do u think it's better move this line before the thread creation?

Yeah, also works

Reply:
We need to set it before `open` returns. Otherwise, if we call another `open` right away before the fetching thread starts, we might get into an illegal state: the status code was `NotInit` but it has already been set by the previous `open` call